src/charm.py

```diff
@@ -801,6 +999,32 @@ def _initialization_secret_set_in_peer_relation(self) -> bool:
             return False
         return False
 
+    def _create_raft_snapshot(self) -> Optional[bytes]:
+        """Creates a snapshot of Vault.
+
+        Returns:
+            bytes: The snapshot content
```

Comment:
Ain't that a lot? Shouldn't we simply create the snapshot and output the file directory?

Reply:
Where would that file be saved?

The idea of this function is to call the `create_snapshot` function of Vault's client, but before doing that we are making sure that vault is initialised and unsealed.

Then we get the content of the snapshot in bytes as returned by vault's client.

Then the caller of this function can save the content to a file if that's intended, or save it into an object on S3 storage as done in this case in the action handler

```
...
         snapshot := self._create_raft_snapshot()
...
         self._upload_byte_content_to_s3(
                ....
                content=snapshot,
                ....
          )
....
```

I improved readability a bit now

Reply:
I think the best approach here would be to return the file-like object (`response.raw`). boto3 should be able to take it as the Body for `put_object()`. Then, the whole file will not be loaded into memory.

Reply:
I see the point @gruyaume 
I fixed it to use the approach proposed by @DanielArndt